File: string_search/brute_force/src/bruteForceSearchDemo.js
// JavaScript
// 文字列の検索: ブルートフォース法 (Brute Force Search)

function search(text, pattern) {
  const n = text.length;
  const m = pattern.length;
  const positions = [];
  
  // パターンが空または検索対象より長い場合
  if (m === 0 || m > n) {
    return positions;
  }
  
  // テキスト内の各位置でパターンとの一致を確認
  for (let i = 0; i <= n - m; i++) {
    let j = 0;
    // パターンの各文字を確認
    while (j < m && text[i + j] === pattern[j]) {
      j++;
    }
    // パターンが完全に一致した場合
    if (j === m) {
      positions.push(i);
    }
  }
  
  return positions;
}

function main() {
  console.log("BruteForceSearch TEST -----> start");

  console.log("\nsearch");
  const input1 = ["ABABCABCABABABD", "ABABD"];
  console.log(`  入力値: ${JSON.stringify(input1)}`);
  const output1 = search(...input1);
  console.log(`  出力値: ${JSON.stringify(output1)}`);

  console.log("\nsearch");
  const input2 = ["AAAAAA", "AA"];
  console.log(`  入力値: ${JSON.stringify(input2)}`);
  const output2 = search(...input2);
  console.log(`  出力値: ${JSON.stringify(output2)}`);

  console.log("\nsearch");
  const input3 = ["ABCDEFG", "XYZ"];
  console.log(`  入力値: ${JSON.stringify(input3)}`);
  const output3 = search(...input3);
  console.log(`  出力値: ${JSON.stringify(output3)}`);

  console.log("\nsearch");
  const input4 = ["ABCABC", "ABC"];
  console.log(`  入力値: ${JSON.stringify(input4)}`);
  const output4 = search(...input4);
  console.log(`  出力値: ${JSON.stringify(output4)}`);

  console.log("\nsearch");
  const input5 = ["ABC", ""];
  console.log(`  入力値: ${JSON.stringify(input5)}`);
  const output5 = search(...input5);
  console.log(`  出力値: ${JSON.stringify(output5)}`);

  console.log("\nBruteForceSearch TEST <----- end");
}

// Node.jsでの実行環境を想定
if (require.main === module) {
  main();
}

File: string_search/brute_force/src/BruteForceSearchDemo.swift
// Swift
// 文字列の検索: ブルートフォース法 (Brute Force Search)

func search(text: String, pattern: String) -> [Int] {
    let n = text.count
    let m = pattern.count
    var positions: [Int] = []
    
    // パターンが空または検索対象より長い場合
    if m == 0 || m > n {
        return positions
    }
    
    // テキスト内の各位置でパターンとの一致を確認
    for i in 0...(n - m) {
        var j = 0
        // パターンの各文字を確認
        while j < m && text[text.index(text.startIndex, offsetBy: i + j)] == pattern[pattern.index(pattern.startIndex, offsetBy: j)] {
            j += 1
        }
        // パターンが完全に一致した場合
        if j == m {
            positions.append(i)
        }
    }
    
    return positions
}

func main() {
    print("BruteForceSearch TEST -----> start")

    print("\nsearch")
    let input1 = ("ABABCABCABABABD", "ABABD")
    print("  入力値: \(input1)")
    let output1 = search(text: input1.0, pattern: input1.1)
    print("  出力値: \(output1)")

    print("\nsearch")
    let input2 = ("AAAAAA", "AA")
    print("  入力値: \(input2)")
    let output2 = search(text: input2.0, pattern: input2.1)
    print("  出力値: \(output2)")

    print("\nsearch")
    let input3 = ("ABCDEFG", "XYZ")
    print("  入力値: \(input3)")
    let output3 = search(text: input3.0, pattern: input3.1)
    print("  出力値: \(output3)")

    print("\nsearch")
    let input4 = ("ABCABC", "ABC")
    print("  入力値: \(input4)")
    let output4 = search(text: input4.0, pattern: input4.1)
    print("  出力値: \(output4)")

    print("\nsearch")
    let input5 = ("ABC", "")
    print("  入力値: \(input5)")
    let output5 = search(text: input5.0, pattern: input5.1)
    print("  出力値: \(output5)")

    print("\nBruteForceSearch TEST <----- end")
}

main()